Fix: Wordwrap for MiniConsole and echo

guhitb opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
* Added a `TBuffer::wrapText`. It takes a QString, and returns a wrapped and indented version of that QString.
* Modified both `TConsole::print` functions to wrap the message before calling `TBuffer::append`.

#### Motivation for adding to Mudlet
MiniConsole output was not being intented. The following script did not work as expected:
```lua
Geyser.MiniConsole:new({name = "indent-test"})
setWindowWrapIndent("indent-test", 10)
echo("indent-test", "This is the song that doesn't end, yes it goes on and on my friend, some people started singing it not knowing what it was, and they'll continue singing it forever just because this is the song that doesn't end...")
```
/claim #5936
fixes https://github.com/Mudlet/Mudlet/issues/5936
#### Other info (issues closed, discussion etc)
Fixed for MiniConsole:
![image](https://github.com/Mudlet/Mudlet/assets/147658676/38841017-1243-4fed-aa86-91cc5dd70e50)

Fixed for echo:
![image](https://github.com/Mudlet/Mudlet/assets/147658676/93b5ed64-b45d-44da-a0d5-7c282a323fec)

guhitb committed:
fixed wrap for miniconsole and echo

guhitb committed:
fixed formatting

guhitb committed:
fixed formatting

vadi2 commented:
Could you also add "fixes https://github.com/Mudlet/Mudlet/issues/5936" to the ticket, that way Github links the issue and PR together.

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2253,6 +2253,64 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+// Applies indentation of mWrapIndent to wrapped lines
+QString TBuffer::wrapText(const QString& text){
+
+    if ( mWrapAt <= mWrapIndent ){
```
Could you apply the clang-tidy formatting here? We have a `.clang-tidy` at the repository root with all the right settings.

guhitb replied:
I ran `clang-tidy -fix src/TBuffer.cpp`, but it's only returning an error. I've tried to fix the formatting manually.

The error:
```
Error while trying to load a compilation database:
Could not auto-detect compilation database for file "src/TBuffer.cpp"
No compilation database found in /home/user/git/forks/Mudlet/src or any parent directory
json-compilation-database: Error while opening JSON database: No such file or directory
fixed-compilation-database: Error while opening fixed database: No such file or directory
Running without flags.
51632 warnings and 1 error generated.
Error while processing /home/user/git/forks/Mudlet/src/TBuffer.cpp.
/home/user/git/forks/Mudlet/src/TTextCodec.h:29:10: error: 'QTextCodec' file not found [clang-diagnostic-error]
#include <QTextCodec>
         ^~~~~~~~~~~~
Suppressed 51632 warnings (51632 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
Found compiler errors, but -fix-errors was not specified.
Fixes have NOT been applied.

Found compiler error(s).
```

guhitb replied:
I got clang-tidy to work by running it from qtcreator.

vadi2 replied:
Sorry, I mentioned the wrong tool. It is `clang-format` that can format the function, if you could apply it to the function you added only.

clang-tidy is still useful though as a static analysis checker!

guhitb replied:
Okay, I pushed `clang-format` changes to the new function.

vadi2 commented:
@demonnic mind testing this one?

guhitb committed:
fixed formatting

guhitb committed:
ran clang-tidy

guhitb committed:
ran clang-format on TBuffer::wrapText

demonnic commented:
So I ran this through the stressinator for 100k lines and the speeds are comparable, but I'm getting extra newlines in the PR version:

## release
![image](https://github.com/Mudlet/Mudlet/assets/3660/73a63122-fd2a-4425-9fba-ba8801447bab)

## PR
![image](https://github.com/Mudlet/Mudlet/assets/3660/3a0e85f5-7c85-4204-86e4-f2f2c9d0bfcb)

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2253,6 +2253,64 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+    int wordsInCurrentLine = 0;
+
+    for (int i = 0; i < text.size(); i++) {
+        bool const at_newline = text.at(i) == '\n';
```
Some `QChar`s are special cased and are available directly (see https://doc.qt.io/qt-5/qchar.html#SpecialCharacter-enum ) so that you do not need to have a C raw-character which would trigger the rejection (at compilation) of raw-string/raw-character that `QT_NO_CAST_FROM_ASCII` / `QT_NO_CAST_TO_ASCII` produce if they are ever `#define`d.

Also, for consistency, put the `const` before the type specifier:
```suggestion
        const bool at_newline = (text.at(i) == QChar::LineFeed);
```

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2253,6 +2253,64 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+        bool const at_newline = text.at(i) == '\n';
+        if (at_newline) {
+            currentLine += currentWord;
+            wrappedText += '\n' + currentLine;
```
Suggestion, use the "QStringBuilder" https://wiki.qt.io/Using_QString_Effectively#QStringBuilder_:_Fast_QString_concatenation to reduce the number of `QString`s that need to be constructed:
```suggestion
            wrappedText += QChar::LineFeed % currentLine;
```

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2253,6 +2253,64 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+
+        currentWord += text.at(i);
+
+        bool const atWordBreak = wordBreaks.indexOf(text.at(i)) > -1;
```
```suggestion
        const bool atWordBreak = (wordBreaks.indexOf(text.at(i)) > -1);
```

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2253,6 +2253,64 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+            wordsInCurrentLine++;
+        }
+
+        int const lineLengthWithWord = currentLine.size() + currentWord.size();
```
Beware, `QString::size()` no longer returns an `int` (well an `int32_t`) type in Qt6, so this may need tweaking to build in the future...
```suggestion
        const qsizetype lineLengthWithWord = currentLine.size() + currentWord.size();
```

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2253,6 +2253,64 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+        }
+
+        int const lineLengthWithWord = currentLine.size() + currentWord.size();
+        bool const needNewLine = lineLengthWithWord >= mWrapAt;
```
```suggestion
        const bool needNewLine = (lineLengthWithWord >= mWrapAt);
```

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2253,6 +2253,64 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+
+        if (needNewLine) {
+            // Current word would cause an overflow, so wrap the text.
+            if (wordsInCurrentLine == 0) {
```
Comparing an integer to zero is more simply put as:
```suggestion
            if (!wordsInCurrentLine) {
```

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2253,6 +2253,64 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+            if (wordsInCurrentLine == 0) {
+                // If a word is too long to fit on a line,
+                // the word will be split between lines
+                int const splitIndex = mWrapAt - currentLine.size();
```
```suggestion
                const qsizetype splitIndex = mWrapAt - currentLine.size();
```

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2253,6 +2253,64 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+                currentLine += currentWord.left(splitIndex);
+                currentWord = currentWord.mid(splitIndex);
+            }
+            wrappedText += '\n' + currentLine;
```
```suggestion
            wrappedText += QChar::LineFeed % currentLine;
```

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2253,6 +2253,64 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+        }
+    }
+    currentLine += currentWord;
+    wrappedText += '\n' + currentLine;
```
```suggestion
    wrappedText += QChar::LineFeed % currentLine;
```

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2253,6 +2253,64 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+        qWarning() << "mWrapAt (" << mWrapAt << ") is too small to accommodate mWrapIndent (" << mWrapIndent << ")";
+    }
+
+    const QString wordBreaks = qsl(",.- ");
```
This methodology, (of looking for these particular characters in the text and considering splitting the line where they are is ***fundamentally*** flawed because it cannot work for locales that do NOT use any of these characters (including spaces) between textual "units" {e.g. Chinese, either Traditional OR Simplified, see https://chinese.stackexchange.com/q/17763 }.

I have long held the opinion that Mudlet needs to instead look at using the `QTextBoundaryFinder` and finding the point in the text where `QTextBoundaryFinder::Line` type boundaries are...

guhitb committed:
moved const before type

guhitb committed:
replaced \n with QChar::LineFeed

guhitb committed:
added parentheses to comparisons

guhitb committed:
replaced int with qsizetype

guhitb committed:
changed wordsInCurrentLine==0 to !wordsInCurrentLine

guhitb committed:
fixed bug, wrapText no longer adds a newline to everything it wraps

guhitb committed:
change variable to camelCase for consistency

guhitb commented:
Implemented the minor changes and removed the newlines added at the end of each `echo`. Text color changes seem to be a second source of newlines, so not 100% yet. I still need to update `wrapText` to use `QTextBoundryFinder`, hopefully that will also fix newlines from color changes.

![image](https://github.com/Mudlet/Mudlet/assets/147658676/32927bce-a0b4-4226-b2de-fb2473ab94da)

guhitb committed:
updated to use QTextBoundryFinder

guhitb committed:
added const and removed extra logic

guhitb commented:
Switched to using `QTextBoundryFinder` and it's working better.

![image](https://github.com/Mudlet/Mudlet/assets/147658676/52e3b2c8-5fef-4591-ac78-c2dcc0bd2aa8)

![image](https://github.com/Mudlet/Mudlet/assets/147658676/ea44e0a7-7958-4a78-9044-38e96d3748fd)

vadi2 commented:
Cool! What is the before & after with 100,000 lines on your machine?

guhitb commented:
Running `stresstest` on 100,000 lines:
- Before: 121 seconds
- After: 124 seconds

![image](https://github.com/Mudlet/Mudlet/assets/147658676/073163be-a523-441b-9d63-b73622f75b32)

![image](https://github.com/Mudlet/Mudlet/assets/147658676/9b9d5834-2cc1-41f8-be0c-8851bcb0f170)

vadi2 commented:
Did you also fix https://github.com/Mudlet/Mudlet/issues/5564 by accident with this?

guhitb commented:
This will wrap Chinese chacters, but the lines will be twice as long due to character width. These two examples were wrapped with the same settings:

![image](https://github.com/Mudlet/Mudlet/assets/147658676/544a8d9b-cccc-4136-a28a-81f0bb5cf4f3)

SlySven commented:
> This will wrap Chinese chacters, but the lines will be twice as long due to character width. ...

As a solution I think that initially this seems good enough to get the prize - I might fiddle with it later on so that it uses the `widechar_width(...)` function that we import from **@ ridiculousFish** 's widechar_width project (that we use elsewhere to work out how wide a grapheme is) to deduce whether the line of text passes the wrapping width in "normal" spaces - so that "wide" graphemes contribute 2 to the total compared to 1 for normal characters...
